includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php

```diff
@@ -1098,6 +1117,213 @@ private function sanitize_disallowed_attribute_values_in_node( DOMElement $node,
+			return false;
+		}
+
+		$input_height = new AMP_CSS_Length(
+			$node->getAttribute( 'height' ),
+			true,
+			true
+		);
+
+		if ( ! $input_height->is_valid() ) {
+			return false;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Validates the layout for the given tag. This involves checking the
+	 * layout, width, height, sizes attributes with AMP specific logic.
+	 *
+	 * @param array[][]  $tag_spec Tag spec list.
+	 * @param DOMElement $node     Tag to validate.
+	 * @return bool True if layout is valid, false if validation fails.
+	 */
+	private function is_valid_layout( $tag_spec, $node ) {
+		$layout_attr = $node->getAttribute( 'layout' );
+
+		$input_width = new AMP_CSS_Length(
+			$node->getAttribute( 'width' ),
+			true,
+			AMP_Rule_Spec::LAYOUT_FLUID === $layout_attr
+		);
+
+		if ( ! $input_width->is_valid() ) {
+			return false;
+		}
+
+		$input_height = new AMP_CSS_Length(
+			$node->getAttribute( 'height' ),
+			true,
+			AMP_Rule_Spec::LAYOUT_FLUID === $layout_attr
+		);
+
+		if ( ! $input_height->is_valid() ) {
+			return false;
+		}
+
+		if ( isset( $tag_spec['amp_layout'] ) ) {
+			$sizes_attr   = $node->getAttribute( 'sizes' );
+			$heights_attr = $node->getAttribute( 'heights' );
+
+			// Now calculate the effective layout attributes.
+			$width  = $this->calculate_width( $tag_spec['amp_layout'], $layout_attr, $input_width );
+			$height = $this->calculate_height( $tag_spec['amp_layout'], $layout_attr, $input_height );
+			$layout = $this->calculate_layout( $layout_attr, $width, $height, $sizes_attr, $heights_attr );
+
+			// Only FLEX_ITEM allows for height to be set to auto.
+			if ( $height->is_auto() && AMP_Rule_Spec::LAYOUT_FLEX_ITEM !== $layout ) {
+				return false;
+			}
+
+			// FIXED, FIXED_HEIGHT, INTRINSIC, RESPONSIVE must have height set.
+			if (
+				(
+					AMP_Rule_Spec::LAYOUT_FIXED === $layout ||
+					AMP_Rule_Spec::LAYOUT_FIXED_HEIGHT === $layout ||
+					AMP_Rule_Spec::LAYOUT_INTRINSIC === $layout ||
+					AMP_Rule_Spec::LAYOUT_RESPONSIVE === $layout
+				) &&
+				! $height->is_set()
+			) {
+				return false;
+			}
+
+			// For FIXED_HEIGHT if width is set it must be auto.
+			if ( AMP_Rule_Spec::LAYOUT_FIXED_HEIGHT === $layout && $width->is_set() && ! $width->is_auto() ) {
+				$node->setAttribute( 'width', 'auto' );
+			}
+
+			// FIXED, INTRINSIC, RESPONSIVE must have width set and not be auto.
+			if (
+				AMP_Rule_Spec::LAYOUT_FIXED === $layout ||
+				AMP_Rule_Spec::LAYOUT_INTRINSIC === $layout ||
+				AMP_Rule_Spec::LAYOUT_RESPONSIVE === $layout
+			) {
+				if ( ! $width->is_set() || $width->is_auto() ) {
+					return false;
+				}
+			}
+
+			// INTRINSIC, RESPONSIVE must have same units for height and width.
+			if (
+				(
+					AMP_Rule_Spec::LAYOUT_INTRINSIC === $layout ||
+					AMP_Rule_Spec::LAYOUT_RESPONSIVE === $layout
+				) &&
+				$width->get_unit() !== $height->get_unit()
+			) {
+				return false;
+			}
+
+			// RESPONSIVE only allows heights attribute.
```

Comment:
This comment seems to be wrong in the JS code already, as the code doesn't do what the comment claims.
```suggestion
			// Height attribute is only allowed for RESPONSIVE layout.
```

I made an upstream PR to that regard: https://github.com/ampproject/amphtml/pull/25581